### PR TITLE
ENH: bump boost-headers-only to 1.80.0

### LINKED ISSUE #15136: ENH: Bump boost.math version
Breaking this out into a separate issue since the [comment](https://github.com/scipy/scipy/pull/13347#issuecomment-983295894) might otherwise easily get lost in a superseded PR:
> @mckib2: It occurs to me that we can try upgrading Boost now that we have a more recent minimum compiler that can handle more modern C++. That may help get upstream fixes for Boost.Math that we've been patching manually, might help with [gh-14901](https://github.com/scipy/scipy/issues/14901) and friends

In particular, now that the follow-up #15066 was merged, it's possible for SciPy to use C++14, which was probably the limiting factor for progressing further with boost?

### PR DESCRIPTION
**Pull request by mckib2**
#### Reference issue

- Closes gh-15136

#### What does this implement/fix?

- upgrades Boost 1.75.0 -> 1.80.0

#### Additional information

- FreeBSD patch may need to be remade

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -3,7 +3,7 @@
 	url = https://github.com/scipy/scipy-mathjax.git
 [submodule "scipy/_lib/boost"]
 	path = scipy/_lib/boost
-	url = https://github.com/scipy/boost-headers-only
+	url = https://github.com/mckib2/boost-headers-only
```
Comment (tupui) [inline .gitmodules:6]: I think this is a leftover from local tests?

Comment (mckib2) [reply .gitmodules:6]: This will need to be changed -- doing some testing to see if we can jump straight to Boost 1.80.0

**Comment by mckib2** [comment]
@h-vetinari CI failure here looks unrelated to Boost changes -- no rush, but if you have a chance to look at impacts to conda-forge and it works, besides the OpenBSD/FreeBSD patch and updating scipy/boost-headers-only, this might be ready to go

**Comment by andyfaff** [comment]
Could you rebase on main? There are several CI fails on macosx_arm64 that I'm hoping my get fixed by this PR. The CI infrastructure isn't present yet in your branch.

**Comment by mckib2** [comment]
> Could you rebase on main? There are several CI fails on macosx_arm64 that I'm hoping my get fixed by this PR. The CI infrastructure isn't present yet in your branch.

Should have latest main now as well as latest upstream patch for M1 compiler bug

**Comment by mdhaber** [comment]
Do I understand that this will fix gh-14901? Any chance it also fixes gh-17101?

**Comment by andyfaff** [comment]
It doesn't fix the issues we're seeing on macosx_arm64 on cirrus.

**Comment by mborland** [comment]
> Do I understand that this will fix [gh-14901](https://github.com/scipy/scipy/issues/14901)? Any chance it also fixes [gh-17101](https://github.com/scipy/scipy/issues/17101)?

These should both be fixed in Boost 1.81.0. 

https://github.com/scipy/scipy/issues/14901 was fixed with https://github.com/boostorg/math/pull/827
https://github.com/scipy/scipy/issues/17101 will be fixed with https://github.com/boostorg/math/pull/846

**Comment by mdhaber** [comment]
It looks like gh-17146 is fixed by https://github.com/boostorg/math/pull/847, too. Thanks @mborland!

Will all these make it in this PR, @mckib2?

**Comment by mborland** [comment]
> It looks like [gh-17146](https://github.com/scipy/scipy/issues/17146) is fixed by [boostorg/math#847](https://github.com/boostorg/math/pull/847), too. Thanks @mborland!
> 
> Do you think we can get 1.81 before SciPy 1.10.0 releases, @mckib2?

Boost 1.81 is scheduled to release the 14 of December. Do you use the standalone version of Boost.Math, or the Boost super project? If you are using the standalone we can cut a 1.80.1 in the event 1.81 doesn't meet your timeline.

**Comment by mckib2** [comment]
We have been generating headers from the "regular" Boost distribution which includes most of the super-project.  I believe we did it this way as we thought we might anticipate using other Boost libraries, but I don't think that's a huge concern for us now. Boost.Math is really all we need.

@mdhaber @rgommers Is it worth the small refactor in scipy/boost-headers-only to use the standalone version of Boost.Math?  It's unclear to me if anyone else is now relying on that repo to provide more than just Boost.Math.

**Comment by mdhaber** [comment]
Anyone else in SciPy? I don't think so. I think you reviewed the only other use of Boost in SciPy that I know of, which is (presumably Math) in `scipy.special`.

**Comment by rgommers** [comment]
> @mdhaber @rgommers Is it worth the small refactor in scipy/boost-headers-only to use the standalone version of Boost.Math? It's unclear to me if anyone else is now relying on that repo to provide more than just Boost.Math.

That does sound like a good idea to me - smaller checkouts, and more flexible in terms of releases it looks like. 

No one else should be relying on that repo, and in case they do they can just fork it and do their own thing. This is meant for SciPy only.

One thing to keep in mind is to not lose the old commits in the `boost-headers-only` repo. It can be a disconnected commit history in the `main` branch though.

**Comment by mdhaber** [comment]
Please revert as much of gh-17272 as possible when this PR gets merged.

**Comment by mckib2** [comment]
If we use the standalone version of Boost.Math, we may no longer need the machinery around the Boost headers-only repo.  In this case, closing this PR in favor of gh-17432 which includes Boost.Math directly

**Comment by mdhaber** [comment]
On the other hand, we've discussed using additional parts of Boost, like the `odeint` and `polygon`. Would these still be available with gh-17432?

**Comment by rgommers** [comment]
It can always be reverted. Boost is huge, so adding a dependency on more of it probably needs a separate conversation. Looking at those topics, it also seems that those may not materialize. `scipy.integrate` gets very little attention, and `polygon` would be for `scipy.spatial` I guess? We also have no owner for QHull there at the moment.

**Pull request closed without merging**